haskell/src/Stablecoin/Client/Parser.hs

```diff
@@ -0,0 +1,522 @@
+        (pure CmdGetPendingContractOwner)
+        "Check if there's a pending contract owner"
+
+    getMasterMinterCmd :: Opt.Mod Opt.CommandFields ClientArgsRaw
+    getMasterMinterCmd =
+      mkCommandParser
+        "get-master-minter"
+        (pure CmdGetMasterMinter)
+        "Get the master minter's address"
+
+    getPauserCmd :: Opt.Mod Opt.CommandFields ClientArgsRaw
+    getPauserCmd =
+      mkCommandParser
+        "get-pauser"
+        (pure CmdGetPauser)
+        "Get the pauser's address"
+
+    getTransferlistCmd :: Opt.Mod Opt.CommandFields ClientArgsRaw
+    getTransferlistCmd =
+      mkCommandParser
+        "get-transferlist"
+        (pure CmdGetTransferlist)
+        "Get the transferlist's address"
+
+    getTotalSupplyCmd :: Opt.Mod Opt.CommandFields ClientArgsRaw
+    getTotalSupplyCmd =
+      mkCommandParser
+        "get-total-supply"
+        (pure CmdGetTotalSupply)
+        "Get the total supply of tokens"
+
+stablecoinClientInfo :: Opt.ParserInfo ClientArgs
+stablecoinClientInfo =
+  Opt.info (Opt.helper <*> versionOption <*> clientArgsParser) $
+  mconcat
+    [ Opt.fullDesc
+    , Opt.progDesc $
+        "Tezos Stablecoin Client: Command line interface for interacting " <>
+        "with the Tezos Stablecoin contract"
+    , Opt.header "Stablecoin Client"
+    , Opt.footerDoc $ usageDoc
+    ]
+
+  where
+    versionOption =
+      Opt.infoOption
+        ("stablecoin-" <> showVersion version)
+        (Opt.long "version" <> Opt.help "Show version.")
+
+    usageDoc :: Maybe Doc
+    usageDoc =
+      Just $ mconcat
+      [ "You can use help for specific COMMAND", linebreak
+      , "EXAMPLE:", linebreak
+      , "  stablecoin-client mint --help", linebreak
+      , "USAGE EXAMPLE:", linebreak
+      , "  stablecoin-client mint --to tz1U1h1YzBJixXmaTgpwDpZnbrYHX3fMSpvb --amount 100500", linebreak
+      , linebreak
+      , "  This command will produce the given amounts of tokens to", linebreak
+      , "  the wallets associated with the given address.", linebreak
+      ]
+
+-- TODO: add this orphan instance to morley
+instance HasCLReader AddressOrAlias where
+  getReader =
+    Opt.str <&> \addrOrAlias ->
+      case parseAddress addrOrAlias of
+        Right addr -> AddressResolved addr
+        Left _ -> AddressAlias addrOrAlias
+  getMetavar = "ADDRESS OR ALIAS"
+
+addressOrAliasOption ::
+  Maybe AddressOrAlias -> "name" :! String -> "help" :! String -> Opt.Parser AddressOrAlias
+addressOrAliasOption = mkCLOptionParser
+
+naturalOption ::
+  Maybe Natural -> "name" :! String -> "help" :! String -> Opt.Parser Natural
+naturalOption = mkCLOptionParser
+
+nonEmptyParser :: Opt.Parser a -> Opt.Parser (NonEmpty a)
+nonEmptyParser p =
+  -- Note: `(:|) <$> p <*> many p` does not work here, because then optparse-applicative
+  -- will display the option's info twice when --help is invoked.
+  -- For this reason, we have to use `some` and the "unsafe" `NonEmpty.fromList`.
```

Comment:
Interesting. It might be a good idea to upstream this function to `optparse-applicative` some time later.

Reply:
Seems like a solution has been very recently added to `optpatse-applicative`, and will be in the next release:

* https://github.com/pcapriotti/optparse-applicative/issues/376
* https://github.com/pcapriotti/optparse-applicative/pull/382

I'll update my comment here to mention this, and that we should use `some1` once it's released.